adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -121,6 +121,121 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+      }, preservesPartitioning = true).zipWithIndex
+      .mapPartitionsWithIndex((idx, iter) => {
+        getBalancedPartitionNumber(iter, partitionMap.get(idx), average)
+      })
+      .partitionBy(new GenomicPositionRangePartitioner(partitions))
+      .mapPartitions(iter => {
+        // sorting an Iterator doesn't seem straight forward, so we cast to a list
+        val listRepresentation = iter.map(_._2).toList.sortBy(_._2.head._1)
+        // we take advantage of this casting to also get the bounds on the partition
+        // and package that into the RDD so additional computation isn't needed
+        Iterator(((listRepresentation.head._1._1, listRepresentation.last._1._2), listRepresentation.map(f => f._2)))
+      }, preservesPartitioning = true)
+
+    replaceRdd(finalPartitionedRDD.flatMap(f => f._2.flatten.map(f => f._2)), Some(finalPartitionedRDD.keys.collect))
+  }
+
+  /**
+   * Infers the correct reference region for each T given a the bounds of the partition (partitionMap)
+   * and returns the iterator with each T paired with the ReferenceRegion that it represents in that
+   * sorted location. This is important because a T can map to multiple ReferenceRegions.
+   *
+   * @param iter The iterator (that partition's data) of T objects
+   * @param partitionMap the bounds of the partition where _._1 is the lower bound and _._2 is the
+   *                     upper bound
+   * @return An iterator of all T's keyed with the specific ReferenceRegion that the T represents in
+   *         the global sort
+   */
+  private def inferCorrectReferenceRegionsForPartition(
+    iter: Iterator[T], partitionMap: (ReferenceRegion, ReferenceRegion)): Iterator[((ReferenceRegion, T))] = {
+
+    // we put the records that map to multiple locations in this ListBuffer
+    // IMPORTANT: these are records that map to multiple locations on this
+    // partition. There should be very few of these
+    // we use this because we have to keep track of which ReferenceRegions
+    // have been used.
+    val duplicates = new ListBuffer[(Seq[ReferenceRegion], T)]()
+    // filter out ReferenceRegions that are not on this partition
+    iter.map(f => (getReferenceRegions(f).filter(g =>
+      g.compareTo(partitionMap._1) >= 0 && g.compareTo(partitionMap._2) <= 0).sorted, f))
+      .map(f => {
+        // the usual case: only one ReferenceRegion for the record
+        if (f._1.length == 1) {
+          (f._1.head, f._2)
+          // if there are multiple, we have to do a little extra work to make
+          // sure that we have the correct ReferenceRegion
+        } else {
+          val index = duplicates.indexOf(f)
+          // if we've already added this record to the duplicates list
+          if (index >= 0) {
+            // get the inferred value from the sorted list of ReferenceRegions
+            val inferredValue = (duplicates(index)._1.head, duplicates(index)._2)
```

Comment:
This variable represents the correct `ReferenceRegion` from the `Seq[ReferenceRegion]` for that GenomicRDD. The use of `infer` here isn't meant to be the same as possible, but more along the lines of deduced. I still think renaming it is a good idea because it's too generic.